torch/lib/c10d/ProcessGroup.hpp

```diff
@@ -43,8 +43,21 @@ class ProcessGroup {
     // If false, the exception function can be called to get details.
     virtual bool isSuccess() const = 0;
 
+    // Ensures that operations on the output tensors that are invoked
+    // after this function returns are correctly sequenced after the
+    // asynchronous completion of this work.
+    virtual void synchronize() = 0;
```

Comment:
Can you add some more docs for this function? Does it synchronize CUDA/CPU/both? Is there a description for the concurrency model of ProcessGroups somewhere? Can you refer to it from here? Would I ever want to call this manually instead of using `wait()`?

Reply:
Yeah, i was gonna ask the same

Reply:
Added:

```
+    // For CUDA tensors, it inserts stream synchronization such that
+    // the streams of the caller wait for completion of the
+    // asynchronous operations on the destination tensors.
+    //
+    // For CPU tensors, it is currently a nop.
+    //
+    // This function should only be used if the caller polls for
+    // completion through the `isCompleted` function, it has returned
+    // true, and the `isSuccess` function also has returned true.
+    //
```